include/zio/zio_fifo_buf.h

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright (c) 2019 Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @file zio_fifo_buf.h
+ *
+ * @brief ZIO FIFO Buffer provides a software zio_buf implementation
+ * based on a zio_fifo
+ */
+
+#ifndef ZEPHYR_INCLUDE_ZIO_FIFO_BUF_H_
+#define ZEPHYR_INCLUDE_ZIO_FIFO_BUF_H_
+
+#include <zio/zio_fifo.h>
+#include <zio/zio_buf.h>
+
+/**
+ * @private
+ * @brief Software FIFO implementation of a zio_buf
+ */
+struct z_zio_fifo_buf {
+	u32_t watermark;
+	u32_t length;
+	struct k_poll_signal signal;
```

Comment:
I definitely considered a semaphore here, but wasn't sure what the comparable cost would be, I don't really think one or the other has any signficant benefits, I was debating hiding most of these struct attributes behind function calls as well to allow for a driver to best decide how to do this stuff rather than exposing these things in a public struct.

Reply:
I've gone ahead and used a semaphore instead